### DISCUSSION TITLE
Windows does not use WSA_FLAG_OVERLAPPED to create sockets, what are the benefits of this?

### DISCUSSION THREAD

**Discussion by number201724** [libuv/libuv#3779] [Q&A]
I see that libuv directly calls the socket function to create a socket on the windows platform, but in Microsoft IOCP example, by default, WSASocket is used to create a socket and set the WSA_FLAG_OVERLAPPED flags. Is there any difference and benefit?

I've seen in various places that some places use the socket function directly, and some places use WSASocket with the WSA_FLAG_OVERLAPPED flag added.

```

      /* Materialize socket and close it */
      socket = WSASocketW(FROM_PROTOCOL_INFO,
                          FROM_PROTOCOL_INFO,
                          FROM_PROTOCOL_INFO,
                          &xfer_queue_item->xfer_info.socket_info,
                          0,
                          WSA_FLAG_OVERLAPPED);
      uv__free(xfer_queue_item);
```
```

static int uv__tcp_try_bind(uv_tcp_t* handle,
                            const struct sockaddr* addr,
                            unsigned int addrlen,
                            unsigned int flags) {
  DWORD err;
  int r;

  if (handle->socket == INVALID_SOCKET) {
    SOCKET sock;

    /* Cannot set IPv6-only mode on non-IPv6 socket. */
    if ((flags & UV_TCP_IPV6ONLY) && addr->sa_family != AF_INET6)
      return ERROR_INVALID_PARAMETER;

    sock = socket(addr->sa_family, SOCK_STREAM, 0);
    if (sock == INVALID_SOCKET) {
      return WSAGetLastError();
    }

    err = uv__tcp_set_socket(handle->loop, handle, sock, addr->sa_family, 0);
    if (err) {
      closesocket(sock);
      return err;
    }
  }
```

**Comment by bnoordhuis** [comment]
I think the difference has to do with the multi-accept mode for listen sockets, where libuv queues multiple accept requests. See the documentation for `uv_tcp_simultaneous_accepts()`.

**Reply by number201724** [reply]
But shouldn't IOCP post multiple accept requests use WSASocket+WSA_FLAG_OVERLAPPED?

**Reply by bnoordhuis** [reply]
I'm not 100% sure but I think that gets you in a situation where the load is unevenly distributed when multiple processes listen on the same socket.
